### PR TITLE
Endpoints for listing entities.

### PR DESCRIPTION
**Pull request by Buwujiu**

### CODE REVIEW

**Comment by Buwujiu** [comment]
Please do not review this pR now, unit tests are failing because this PR depends on PR#173. Tests should pass once 173 is merged. Moving back to in progress.

File: src/main/java/org/icgc/argo/program_service/utils/EntityService.java [hunk 1]
```diff
@@ -6,11 +6,14 @@
 import com.google.common.collect.ImmutableSet;
```
Comment (d8660091) [inline src/main/java/org/icgc/argo/program_service/utils/EntityService.java]: Name of this file is kind of misleading, as *Service files are inside services folder.

File: src/main/java/org/icgc/argo/program_service/services/ProgramService.java [hunk 2]
```diff
@@ -314,4 +314,37 @@ public void removeProgram(String name) throws EmptyResultDataAccessException {
                 .listAll());
     return List.copyOf(new LinkedHashSet<ProgramEntity>(programs));
   }
+
+  public List<CancerEntity> listCancers() {
+    val cancers = cancerRepository.findAll();
+    return List.copyOf(new LinkedHashSet<CancerEntity>(cancers));
```
Comment (d8660091) [inline src/main/java/org/icgc/argo/program_service/services/ProgramService.java]: Can it be simplified to ```
return cancerRepository.findAll();
``` ?

File: src/main/java/org/icgc/argo/program_service/grpc/ProgramServiceImpl.java [hunk 3]
```diff
@@ -261,10 +263,66 @@ public void removeProgram(RemoveProgramRequest request, StreamObserver<Empty> re
...
+  public void addInstitutions(AddInstitutionsRequest request, StreamObserver<AddInstitutionsResponse> responseObserver) {
+    val names = request.getNamesList().stream().map(name -> commonConverter.unboxStringValue(name)).collect(toImmutableList());
+    try {
+      val institutions = programService.addInstitutions(names);
+      val response = programConverter.institutionsToAddInstitutionsResponse(institutions);
+      responseObserver.onNext(response);
+      responseObserver.onCompleted();
+    } catch (DataIntegrityViolationException e){
+      log.error("Exception throw in addInstitutions: {}", e.getMessage());
+      e.printStackTrace();
```
Comment (d8660091) [inline src/main/java/org/icgc/argo/program_service/grpc/ProgramServiceImpl.java]: This prints to standard error stream. I think log.error(Message, Throwable) should be able to handle the stack trace.
Reference:
https://www.slf4j.org/api/src-html/org/slf4j/Logger.html#line.723

File: src/main/java/org/icgc/argo/program_service/converter/ProgramConverter.java [hunk 4]
```diff
@@ -189,6 +312,30 @@ default ListProgramsResponse programEntitiesToListProgramsResponse(Collection<Pr
     return programEntitiesToListProgramsResponse(0, programEntities);
   }
 
+  default ListCancersResponse cancerEntitiesToListCancersResponse(Collection<CancerEntity> cancerEntities){
+    return cancerEntitiesToListCancersResponse(0, cancerEntities);
```
Comment (d8660091) [inline src/main/java/org/icgc/argo/program_service/converter/ProgramConverter.java]: We can save 11 lines of code
```
ListCancersResponse.newBuilder().addAllCancers(cancerEntities.stream().map(this::cancerEntityToCancer).collect(Collectors.toList()));
```

File: src/main/java/org/icgc/argo/program_service/services/ProgramService.java (hunk 2 above)
Comment (andricDu) [reply src/main/java/org/icgc/argo/program_service/services/ProgramService.java]: `List.copyOf`will make it immutable which is generally a good thing.

You can simplify the whole method to:

```java
  public List<CancerEntity> listCancers() {
    return List.copyOf(cancerRepository.findAll());
  }
```

File: src/main/java/org/icgc/argo/program_service/services/ProgramService.java [hunk 5]
```diff
@@ -314,4 +314,37 @@ public void removeProgram(String name) throws EmptyResultDataAccessException {
...
+  }
+
+  public List<PrimarySiteEntity> listPrimarySites() {
+    val sites = primarySiteRepository.findAll();
+    return List.copyOf(new LinkedHashSet<PrimarySiteEntity>(sites));
```
Comment (andricDu) [inline src/main/java/org/icgc/argo/program_service/services/ProgramService.java]: can probably simplify all of these.

File: src/main/java/org/icgc/argo/program_service/utils/EntityService.java (hunk 1 above)
Comment (Buwujiu) [reply src/main/java/org/icgc/argo/program_service/utils/EntityService.java]: It's under utils, so I think it's fine.

**Pull request merged**